cmem/encoder.go

```diff
@@ -0,0 +1,170 @@
+// Copyright ©2018 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package cmem
+
+import (
+	"encoding/binary"
+	"fmt"
+	"math"
+	"reflect"
+	"unsafe"
+)
+
+var nativeEndian binary.ByteOrder
+
+func init() {
+	one := uint16(0x1)
+	nativeEndian = [...]binary.ByteOrder{
+		binary.BigEndian,
+		binary.LittleEndian,
+	}[(*[2]byte)(unsafe.Pointer(&one))[0]]
+}
+
+// Encoder is a wrapper type for information necessary to create and
+// subsequently write an in memory object to a PacketTable using Append().
+type Encoder struct {
+	// Buf contains the encoded data.
+	Buf    []byte
+	offset int
+}
+
+// Encode encodes the value passed as data to binary form stored in []Buf. This
+// buffer is a Go representation of a C in-memory object that can be appended
+// to e.g. a HDF5 PacketTable.
+//
+// Struct values must only have exported fields, otherwise Encode will panic.
+func (enc *Encoder) Encode(data interface{}) error {
+	padding := enc.offset - len(enc.Buf)
+
+	if padding > 0 {
+		enc.Buf = append(enc.Buf, make([]byte, padding)...)
+	}
+
+	if data, ok := data.(CMarshaler); ok {
+		raw, err := data.MarshalC()
+		if err != nil {
+			return err
+		}
+
+		enc.Buf = append(enc.Buf, raw...)
+
+		return nil
+	}
+
+	rv := reflect.Indirect(reflect.ValueOf(data))
+	rt := rv.Type()
+
+	switch rt.Kind() {
+	case reflect.Slice, reflect.Array:
+		for i := 0; i < rv.Len(); i++ {
+			if err := enc.Encode(rv.Index(i).Interface()); err != nil {
+				return err
+			}
+		}
+		return nil
+
+	case reflect.Struct:
+		offset := enc.offset
+		for i := 0; i < rv.NumField(); i++ {
+			sfv := rv.Field(i).Interface()
+			// In order to keep the memory offset always correct, we use the
+			// structs offset.
+			enc.offset = offset + int(rt.Field(i).Offset)
+			if err := enc.Encode(sfv); err != nil {
+				return err
+			}
+			// Reset the offset to the correct array size.
+			enc.offset = offset + int(rt.Size())
+		}
+		return nil
+
+	case reflect.String:
+		str := append([]byte(rv.String()), 0)
+		strPtr := uintptr(unsafe.Pointer(&str[0]))
+
+		var raw [unsafe.Sizeof(strPtr)]byte
```

Comment:
As another note. Given that hdf5 "does all the wrong things here", we can actually write this whole thing as:

```
// This direct machine conversion is only used
// because HDF5 uses machine endianism.
//
// DO NOT DO THIS AT HOME.
//
raw := (*[unsafe.Sizeof(strPtr)]byte)(strPtr)
enc.Buf = append(enc.Buf, raw[:]...)
enc.offset += len(raw)
```

This is horrific, but in this context, given that they already are already contributing to the net global pain, this is probably the best approach (the comment should be included verbatim).

Reply:
Not just the comment, but the three lines below as well. If we're going to be bad, let's do it properly.

Reply:
Actually... That doesn't work. It will yield in:
`./encoder.go:99: cannot convert strPtr (type uintptr) to type *[8]byte`

Reply:
I left some of it out :|

`raw := (*[unsafe.Sizeof(strPtr)]byte)(unsafe.Pointer(strPtr))`

Reply:
And then continued to be incorrect - edited above.

Reply:
I tried myself the `unsafe.Pointer(strPtr)` version before, which resulted in a seg violation.
Unfortunately, I have to tell you that your version does the same... :disappointed:

Reply:
I've confirmed that this `raw := (*[unsafe.Sizeof(uintptr(0))]byte)(unsafe.Pointer(&str))` works. Sorry for the bad advice previously.

Reply:
Yes! It does. No problemos :)